### DISCUSSION TITLE
Bringing in own expected return and expected covariance

### DISCUSSION THREAD

**Discussion by omroot** [skfolio/skfolio#66] [Q&A] [answered]
Hi, 

I am trying to use the framework to try the variety of optimization techniques but using my own expected excess returns and my own modeled covariance matrix , but it doesn't seem there is an obvious way on how to do it. 

Any help would be much appreciated ... 

Thanks ! 
Best

**Comment by HugoDelatte** [comment, marked as answer]
Hello,

Yes it's possible, an example is given in the user guide [here](https://skfolio.org/user_guide/optimization.html#custom-estimator)
Let me know if it solves your problem.

**Reply by omroot** [reply]
Thank you, I think I got ... This works for me:


` import numpy as np

from skfolio.moments import BaseMu
from skfolio.moments import BaseCovariance
from skfolio.optimization import MeanRisk
from skfolio.preprocessing import prices_to_returns
from skfolio.prior import EmpiricalPrior


my_covariance = np.array([[0.40755159, 0.03175842, 0.05183923, 0.05663904, 0.0330226 ,
                            0.00827775, 0.02165938, 0.01332419, 0.0343476 , 0.02249903],
                           [0.03175842, 0.9063047 , 0.03136385, 0.02687256, 0.01917172,
                            0.00934384, 0.02495043, 0.00761036, 0.02874874, 0.01336866],
                           [0.05183923, 0.03136385, 0.19490901, 0.04408485, 0.03006772,
                            0.01322738, 0.03525971, 0.0115493 , 0.0427563 , 0.02057303],
                           [0.05663904, 0.02687256, 0.04408485, 0.19528471, 0.02777345,
                            0.00526665, 0.01375808, 0.00780878, 0.02914176, 0.01640377],
                           [0.0330226 , 0.01917172, 0.03006772, 0.02777345, 0.34059105,
                            0.00777055, 0.02067844, 0.00736409, 0.02542657, 0.01284075],
                           [0.00827775, 0.00934384, 0.01322738, 0.00526665, 0.00777055,
                            0.15983874, 0.02105575, 0.00518686, 0.01723737, 0.00723779],
                           [0.02165938, 0.02495043, 0.03525971, 0.01375808, 0.02067844,
                            0.02105575, 0.68056711, 0.01377882, 0.04627027, 0.01926088],
                           [0.01332419, 0.00761036, 0.0115493 , 0.00780878, 0.00736409,
                            0.00518686, 0.01377882, 0.95526918, 0.0106553 , 0.00760955],
                           [0.0343476 , 0.02874874, 0.0427563 , 0.02914176, 0.02542657,
                            0.01723737, 0.04627027, 0.0106553 , 0.31681584, 0.01854318],
                           [0.02249903, 0.01336866, 0.02057303, 0.01640377, 0.01284075,
                            0.00723779, 0.01926088, 0.00760955, 0.01854318, 0.11079287]])
my_expected_return = np.array([1.175, 1.19 , 0.396, 1.12 , 0.346, 0.679, 0.089, 0.73 , 0.481,1.08 ])

class MyCustomMu(BaseMu):
    def __init__(self):
        super().__init__() 
    def fit(self, X, y):
        self.mu_ = y
        return self
    

class MyCustomCovariance(BaseCovariance):
    def __init__(self):
        super().__init__()
    def fit(self, X, y):
        self._set_covariance(X)
        return self


model = MeanRisk(
    prior_estimator=EmpiricalPrior(mu_estimator = MyCustomMu(),
                                   covariance_estimator=MyCustomCovariance()),
)
model.fit(X=my_covariance, y=my_expected_return)

print(model.weights_)`

------
[0.03696819 0.02690061 0.09494258 0.12577619 0.07674607 0.21935595
 0.02998683 0.03596317 0.06134975 0.29201067]

**Reply by HugoDelatte** [reply]
The problem with your implementation is that it's breaking the API: `predict`, cross validation, pipelines, hyperparameters tuning, risk measures based on the returns series won't work. 
`X` should be the assets returns and the moment estimation should be computed inside the estimator classes otherwise you will have data leakage when doing train/test or cross validation. If you don't know the moments model (coming from an external source for example), you can pass them using [Metadata  Routing](https://skfolio.org/user_guide/metadata_routing.html) instead of `X` and `y`.

**Reply by temph2020** [reply]
> [Metadata Routing](https://skfolio.org/user_guide/metadata_routing.html)

I have already calculated the covariance matrix cov_mat externally, and I want to pass this matrix directly to skfolio. I checked the ImpliedCovariance class, which accepts an external parameter implied_vol. I tried to mimic it by creating my class MyCustomCovariance that accepts an external parameter cov_mat, but I failed. My code is as follows, and I don’t know how to correct it.


`def get_skfolio_weight_by_cov(alpha, returns, cov_mat):    
    from sklearn import set_config
    from skfolio import  RiskMeasure   
    from skfolio.optimization import MeanRisk, ObjectiveFunction
    # from skfolio.preprocessing import prices_to_returns
    # from skfolio.moments import DenoiseCovariance, ShrunkMu
    from skfolio.prior import EmpiricalPrior
    from skfolio.moments import ImpliedCovariance, BaseCovariance
    import sklearn.utils.metadata_routing as skm
    from skfolio.utils.tools import check_estimator
    
    
    set_config(enable_metadata_routing=True)
    
    def my_expected_return(w):
        return alpha @ w
    
    class MyCustomCovariance(BaseCovariance):
        def __init__(self):
            super().__init__()


        def fit(self, X, cov_mat=None, **fit_params):
            X = self._validate_data(X)
            if cov_mat is not None:
                # noinspection PyTypeChecker
                fit_params["cov_mat"] = cov_mat

            routed_params = skm.process_routing(self, "fit", **fit_params)

            # fitting estimators
            self.prior_covariance_estimator_ = check_estimator(
                self.prior_covariance_estimator,
                default=EmpiricalCovariance(),
                check_type=BaseCovariance,
            )
            # noinspection PyArgumentList
            self.prior_covariance_estimator_.fit(
                X, y, **routed_params.prior_covariance_estimator.fit
            )
     
            self._set_covariance(cov_mat)
            return self
    
    model = MeanRisk(
    risk_measure=RiskMeasure.VARIANCE,
    objective_function=ObjectiveFunction.MAXIMIZE_RATIO,
    overwrite_expected_return=my_expected_return,
    
    prior_estimator=EmpiricalPrior(
        covariance_estimator=MyCustomCovariance(
        ).set_fit_request(cov_mat=True)), 
    



    solver = 'MOSEK',
    save_problem = True,
    portfolio_params=dict(name="Max Sharpe"),
    
    
    )
    model.fit(returns, cov_mat=cov_mat)
    print('solving status',model.problem_.status)

    print('sum weight', sum(model.weights_))
    


    return model.weights_`

**Reply by HugoDelatte** [reply]
Your routed metadata should have same dimension as your input returns `X`.   In the `ImpliedCovariance` example, the implied_vols are the implied volatilities time series for each assets, not the covariance matrix. Otherwise you will get data leakage as soon as you perform cross-validation or any other resampling. 
Your implementation of your custom cov matrix should live inside your custom Covariance estimator. 

Otherwise if you really need to do that you can simply pass your external cov in your estimator init params (but it's not recommended for the reasons explained above)
